src/timemachine.c

```diff
@@ -0,0 +1,613 @@
+ *
+ * \author Mat Oldham <mat.oldham@gmail.com>
+ *
+ * In-memory PCAP buffer for retrieving previously seen packets after an alert 
+ * fires
+ */
+
+#include "suricata-common.h"
+#include "debug.h"
+#include "decode.h"
+#include "detect.h"
+#include "conf.h"
+#include "output.h"
+
+#include "threads.h"
+#include "threadvars.h"
+#include "tm-threads.h"
+
+#include "util-error.h"
+#include "util-debug.h"
+#include "util-time.h"
+#include "util-byte.h"
+#include "util-misc.h"
+#include "util-cpu.h"
+#include "util-atomic.h"
+#include "util-time.h"
+
+#include "timemachine.h"
+#include "timemachine-flow.h"
+#include "timemachine-heap.h"
+#include "timemachine-output.h"
+#include "timemachine-packet.h"
+
+#define MODULE_NAME                               "TimeMachine"
+#define DEFAULT_MAX_MEMORY                        1024 * 1024 * 1024
+#define DEFAULT_MAX_PACKET_SIZE                   1500
+#define DEFAULT_HEAP_PREALLOC_COUNT               5000
+#define DEFAULT_HEAP_EXPAND_BY                    1000
+#define DEFAULT_OUTPUT_TIMEOUT                    21600
+
+SC_ATOMIC_DECLARE(uint32_t, thread_cnt);
+
+/* global pcap data for when we're using multi mode. At exit we'll
+ * merge counters into this one and then report counters. */
+static TimeMachineData *g_tm_data = NULL;
+
+static TmEcode TimeMachineProcess(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueue *);
+static TmEcode TimeMachineInit(ThreadVars *, void *, void **);
+static TmEcode TimeMachineDeinit(ThreadVars *, void *);
+static OutputCtx *TimeMachineInitCtx(ConfNode *);
+static void TimeMachineDeInitCtx(OutputCtx *);
+
+void TmModuleTimeMachineRegister(void)
+{
+    tmm_modules[TMM_TIMEMACHINE].name = MODULE_NAME;
+    tmm_modules[TMM_TIMEMACHINE].ThreadInit = TimeMachineInit;
+    tmm_modules[TMM_TIMEMACHINE].Func = TimeMachineProcess;
+    tmm_modules[TMM_TIMEMACHINE].ThreadDeinit = TimeMachineDeinit;
+    tmm_modules[TMM_TIMEMACHINE].RegisterTests = NULL;
+
+    OutputRegisterModule(MODULE_NAME, "timemachine", TimeMachineInitCtx);
+
+    SC_ATOMIC_INIT(thread_cnt);
+    return;
+}
+
+/**
+ * \brief TimeMachine main logging function
+ *
+ * \param t threadvar
+ * \param p packet
+ * \param data thread module specific data
+ * \param pq pre-packet-queue
+ * \param postpq post-packet-queue
+ *
+ * \retval TM_ECODE_OK on succes
+ * \retval TM_ECODE_FAILED on serious error
+ */
+static TmEcode TimeMachineProcess(ThreadVars *t, Packet *p, void *thread_data, PacketQueue *pq,
+                 PacketQueue *postpq)
+{    
+    TimeMachineFlow *flow;
+    TimeMachineHeap *heap;
+    TimeMachineMemPool *mem_pool;
+    TimeMachinePacket *packet;
+    struct pcap_pkthdr pkthdr;
+    struct timeval current_time;
+    
+    TimeGet(&current_time);
+    
+    TimeMachineThreadData *td = (TimeMachineThreadData *)thread_data;
+    TimeMachineData *tm = td->tm_data;
+    
+    /* make sure packet has real data */
+    if (GET_PKT_PAYLOAD_LEN(p) == 0) {
+        return TM_ECODE_OK;
+    }
+    
```

Comment:
Yes, that's by design.  We could add the handshake though at the cost of
the amount of packets that can be saved.  For very short lived connections,
this can cause a lot of pollution within the heap that I didn't see the
value in having and ultimately takes up space from the heap of more
interesting traffic.

On Tue, Oct 20, 2015 at 8:33 AM, Jason Ish <notifications@github.com> wrote:

> In src/timemachine.c
> <https://github.com/inliniac/suricata/pull/1663#discussion_r42511590>:
>
> > +    TimeMachineHeap *heap;
> > +    TimeMachineMemPool *mem_pool;
> > +    TimeMachinePacket *packet;
> > +    struct pcap_pkthdr pkthdr;
> > +    struct timeval current_time;
> > +
> > +    TimeGet(&current_time);
> > +
> > +    TimeMachineThreadData *td = (TimeMachineThreadData *)thread_data;
> > +    TimeMachineData *tm = td->tm_data;
> > +
> > +    /* make sure packet has real data */
> > +    if (GET_PKT_PAYLOAD_LEN(p) == 0) {
> > +        return TM_ECODE_OK;
> > +    }
> > +
>
> In TCP, doesn't this suppress the handshake setup? I did run it for a
> while, and noticed my output pcaps were missing session setup. Is that by
> design?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/inliniac/suricata/pull/1663/files#r42511590>.
>